OpenSimRT/RealTime/experimental/RealTimeAnalysisExtended/src/RealTimeAnalysisExtended.cpp

```diff
@@ -0,0 +1,192 @@
+            auto pose = inverseKinematics->solve(acquisitionData.IkFrame);
+
+            // filter ik results
+            UnfilteredData unfilteredData;
+            unfilteredData.t = pose.t;
+            unfilteredData.q = pose.q;
+            unfilteredData.externalWrenches = acquisitionData.ExternalWrenches;
+            auto filteredData =
+                    lowPassFilter->filter({pose.t, unfilteredData.toVector()});
+
+            // push to buffer when filter is ready
+            if (!filteredData.isValid) continue;
+            buffer.add(filteredData);
+        }
+    } catch (exception& e) {
+        cout << e.what() << endl;
+
+        // raise termination flag
+        terminationFlag = true;
+
+        // notify buffer to not get stuck in processing thread
+        buffer.externalNotify();
+    }
+}
+
+void RealTimeAnalysisExtended::processing() {
+    try {
+        FilteredData filteredData;
+        Vector am, fm, residuals, reactionWrenchVector;
+        Vector_<SpatialVec> reactionWrenches;
+        while (true) {
+            if (shouldTerminate()) THROW_EXCEPTION("Processing terminated.");
+
+            // get data from buffer
+            auto data = buffer.get(1)[0];
+            filteredData.fromVector(data.t, data.x, data.xDot, data.xDDot,
+                                    model.getNumCoordinates());
+
+            // grfm prediction
+            std::vector<ExternalWrench::Input> externalWrenches;
+            if (parameters.useGRFMPrediction) {
+                // update detector
+                if (parameters.detectorUpdateMethod ==
+                    PhaseDetectorUpdateMethod::INTERNAL)
+                    parameters.internalPhaseDetectorUpdateFunction(
+                            data.t, data.x, data.xDot, data.xDDot);
+                else if (parameters.detectorUpdateMethod ==
+                         PhaseDetectorUpdateMethod::EXTERNAL)
+                    parameters.externalPhaseDetectorUpdateFunction();
+                else
+                    THROW_EXCEPTION("Wrong detector update method");
+
+                // solve grfm prediction
+                auto grfmOutput = grfmPrediction->solve(
+                        {filteredData.t, filteredData.q, filteredData.qd,
+                         filteredData.qdd});
+
+                // setup wrenches
+                ExternalWrench::Input grfRightWrench = {
+                        grfmOutput.right.point, grfmOutput.right.force,
+                        grfmOutput.right.torque};
+                ExternalWrench::Input grfLeftWrench = {grfmOutput.left.point,
+                                                       grfmOutput.left.force,
+                                                       grfmOutput.left.torque};
+                externalWrenches = {grfRightWrench, grfLeftWrench};
+            } else {
+                // use filtered external measured wrenches
+                externalWrenches = filteredData.externalWrenches;
```

Comment:
In the constructor: ensure that useGRFMPrediction = false when external wrenches are provided.

Reply:
I already check if the `parameters.useGRFMPrediction` is true/false. I don't see how I can check in the constructor if external wrenches are provided. The `wrenchParameters` can be used for JR as well, so there is no point of check it.